woocommerce/woocommerce-gutenberg-products-block
# Show decimals in mini cart regular price would not have decimals

Fixes #5018 & #5019

### Problem

In #5019, @dinhtungdu reported that the decimals in the mini cart block are not visible, when the price is 0. After viewing the unit tests of this fix, I noticed that this problem not only affects a $0 price, but also a price that has no decimals, such as $15.00. This price will then appear as $15 in the mini cart.

### Reason

To format the price, we convert the regular price into a decimal price by multiplying the regular price with 10 to the power of the number of decimals. Let's say the merchant sets the number of decimals to 2 and the cart contains products with a value of $23.05. We'd then convert the regular price to a decimal price as follows:

```
23.05 * 10 ^ 2 = 2305
```

Now, let's take a look what happens when we convert a $0 price:

```
0 * 10 ^ 2 = 0
```

To keep the decimal zeros, we'd expect the decimal price `000`, but we get the decimal price `0`. 

Let's also take a quick look what happens when we convert a $15 price:

```
15.00 * 10 ^ 2 = 1500
```

So far, this is ok. However, when formatting the price in https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/6cb995c49f11ee59f03951a57ab8a93b6de01753/packages/prices/utils/price.ts#L137, the decimals get lost:

```
1500 / 10 ^ 2 = 15
```

### Solution

To solve this problem, I'm splitting the price into an integer and a decimal part. That way, I can ensure that the decimals do not get lost independent if the cart holds a $0 price or a higher price, but without decimals, e.g. $15.

### Testing

#### Automated testing

1. Run the following command in the terminal:
```
npm run test -i packages/prices/utils/test/price.js
```

#### Manual testing

##### Test $0.00 value

1. Go to `WP Admin » WooCommerce » Setting » General » Currency options`.
2. Set the number of decimals to `2`.
3. Create a test page with the mini cart block and save it.
4. Open the mini cart in the frontend.
5. See that the initial price shows as `$0.00`.
6. Hover over the price and see that it remains `$0.00`.

<img src="https://user-images.githubusercontent.com/3323310/142971878-4e74878e-260c-427e-ad7d-18bf87e66559.png" width="500">

##### Test number of decimals

1. Go to `WP Admin » WooCommerce » Setting » General » Currency options`.
2. Set the number of decimals to `3`.
3. Open the mini cart in the frontend.
4. See that the initial price shows as `$0.000`.
5. Hover over the price and see that it remains `$0.000`.

<img src="https://user-images.githubusercontent.com/3323310/142971888-cf57087d-fc6c-46ad-8839-bfaaef1bba60.png" width="500">

##### Test separators

1. Go to `WP Admin » WooCommerce » Setting » General » Currency options`.
2. Set the thousand separator to `.` and the decimal separator to `,`.
3. Go to the frontend and add a product to the cart.
4. Increase the order quantity so that the price is above $1.000.
5. See that the thousand separator as `.` and the decimal separator as `,`.
6. Hover over the price and see that the thousand separator and the decimal separator remain the same.

<img src="https://user-images.githubusercontent.com/3323310/142971901-ee783f5f-0879-44b1-90f9-a4825e6f97cf.png" width="500">

##### Test currency and currency position

1. Go to `WP Admin » WooCommerce » Setting » General » Currency options`.
2. Set the currency to `Euro (€)` and the currency position to `Right with space`.
3. Go to the frontend and add a product to the cart.
4. See that the currency is `Euro (€)` and that it's place on the right-hand side, including a space.
5. Hover over the price and the currency and the currency position remain the same.

<img src="https://user-images.githubusercontent.com/3323310/142971382-9a43e702-a5e9-4eca-9544-cb663f59e21e.png" width="500">

### Changelog

> Ensure that formated price shows the expected number of decimals, thousand separator and decimal separator.
